pkg/gui/controllers/files_controller.go

```diff
@@ -748,6 +755,90 @@ func (self *FilesController) createStashMenu() error {
 	})
 }
 
+func (self *FilesController) openCopyMenu() error {
+	copyNameNoItemSelectionErr := utils.ResolvePlaceholderString(
+		self.c.Tr.FileCopyNoItemSelectedError,
+		map[string]string{"target": "name"},
+	)
+	copyNameItem := &types.MenuItem{
+		Label: self.c.Tr.CopyFileName,
+		OnPress: func() error {
+			name := self.context().GetSelectedItemId()
+			if name == "" {
+				return self.c.ErrorMsg(copyNameNoItemSelectionErr)
+			}
+
+			return self.copyToClipboard(name)
+		},
+		Key: 'n',
+	}
+
+	copyDiffNoItemSelectionErr := utils.ResolvePlaceholderString(
+		self.c.Tr.FileCopyNoItemSelectedError,
+		map[string]string{"target": "diff"},
+	)
+	copyFileDiffItem := &types.MenuItem{
+		Label:   self.c.Tr.CopySelectedDiff,
+		Tooltip: self.c.Tr.CopyFileDiffTooltip,
+		OnPress: func() error {
+			path := self.context().GetSelectedPath()
+			if path == "" {
+				return self.c.ErrorMsg(copyDiffNoItemSelectionErr)
+			}
+			diff, err := self.c.Git().Diff.GetPathDiff(path, self.hasPathStagedChanges())
+			if err != nil {
+				return self.c.Error(err)
+			}
+			return self.copyToClipboard(diff)
+		},
+		Key: 's',
+	}
+
+	if self.context().GetSelected() == nil {
+		copyNameItem.DisabledReason = copyNameNoItemSelectionErr
+		copyFileDiffItem.DisabledReason = copyDiffNoItemSelectionErr
+	}
+
+	return self.c.Menu(types.CreateMenuOptions{
+		Title: self.c.Tr.CopyToClipboardMenu,
+		Items: []*types.MenuItem{
+			copyNameItem,
+			copyFileDiffItem,
+			{
+				Label:   self.c.Tr.CopyAllFilesDiff,
+				Tooltip: self.c.Tr.CopyFileDiffTooltip,
+				OnPress: func() error {
+					diff, err := self.c.Git().Diff.GetAllDiff(self.c.Helpers().WorkingTree.AnyStagedFiles())
+					if err != nil {
+						return self.c.Error(err)
+					}
+					return self.copyToClipboard(diff)
+				},
+				Key: 'a',
+			},
+		},
+	})
+}
+
+func (self *FilesController) hasPathStagedChanges() bool {
+	node := self.context().GetSelected()
+	return node.SomeFile(func(t *models.File) bool {
+		return t.HasStagedChanges
+	})
+}
+
+func (self *FilesController) copyToClipboard(content string) error {
+	if content == "" {
```

Comment:
I would be happier if we could check this in advance too, and disable the commands with an appropriate disabledReason, instead of having to check afterwards.

As far as I can tell, this can only happen when you copy the diff for a file, and the file is untracked (or for a directory which has only untracked children), or when you copy the diff for all files, and there aren't any (or only untracked ones). It should be possible to check for all of this up front. Or am I missing some other case?

Reply:
No, you're not missing any case (as far as I know). I thought it would have been safer/simpler to implement this as is, but it should be reasonably simple to check for all the cases that return an empty string and turn off the action in advance. If we're missing some cases, we'll see them showing in the issue section, and it'll be straightforward to patch them.